[Docs] AS Integrations AWS Lambda 2.0.0 Update (#7329)
apollographql/apollo-server: docs/source/deployment/lambda.mdx

 serverless remove
```

## Middleware

In order to implement event and result mutations, type-safe middleware can be passed to the `startServerAndCreateLambdaHandler` call. The API is as follows:

```ts

import { middleware, startServerAndCreateLambdaHandler, handlers } from "@as-integrations/aws-lambda";

const requestHandler = handlers.createAPIGatewayProxyEventV2RequestHandler()

> // All middleware must be async and will be strictly typed based on the `requestHandler`'s types
> const middlewareFn: middleware.MiddlewareFn<typeof requestHandler> = async (event) => {
>   // event is intended to be mutable and can be updated here
> 
>   // optional result mutation, otherwise function returns `Promise<void>`
>   return async (result) => {
>     // result is intended to be mutable and can be updated here
> 
>     // `Promise<void>` return
>   }
> }

Reviewer: I don't have a great suggestion here besides maybe just "don't say anything", but suggesting returning `Promise<void>` is a little weird and possibly confusing when they can return nothing or implicitly return, though I agree that it's technically correct.

Suggesting some changes, but feel free to tweak them.
```suggestion
// Middleware is an async function whose type is based on your request handler. Middleware
// can read and mutate the incoming event. Additionally, returning an async function from your
// middleware allows you to read and mutate the result before it's sent.
const middlewareFn: middleware.MiddlewareFn<typeof requestHandler> = async (event) => {
  // read or update the event here

  // optionally return a callback to access the result
  return async (result) => {
    // read or update the result here
  }
}
```

> startServerAndCreateLambdaHandler(..., ..., {

Reviewer: Eventually we'll be able to prettify code blocks in our docs (https://github.com/apollographql/apollo-server/pull/6914) so I'd like to have valid syntax wherever possible.
```suggestion
startServerAndCreateLambdaHandler(server, requestHandler, {
```

  middleware: [middlewareFn],
})
```

> An example use case would be for cookie modification. The `APIGatewayProxyStructuredResultV2` type contains a property `cookies` that then mutate to multiple `Set-Cookie` headers.

Reviewer: ```suggestion
One use case for middleware is cookie modification. The `APIGatewayProxyStructuredResultV2` type contains a property `cookies` which can be pushed to. This allows you to set multiple `set-cookie` headers in the response.
```

```ts
import {
  startServerAndCreateLambdaHandler,
  middleware,
  handlers,
} from '@as-integrations/aws-lambda';
import { server } from './server';
import { refreshCookie } from './cookies';

const requestHandler = handlers.createAPIGatewayProxyEventV2RequestHandler();

// Utilizing typeof
const cookieMiddleware: middleware.MiddlewareFn<typeof requestHandler> = async (
  event,
) => {
  // Access existing cookies and produce a refreshed one
  const cookie = refreshCookie(event.cookies);
  return async (result) => {

>     // Assure proper initialization of the cookies property on the result

Reviewer: ```suggestion
    // Ensure proper initialization of the cookies property on the result
```

    result.cookies = result.cookies ?? [];
    // Result is mutable so it can be updated here
    result.cookies.push(cookie);
  };
};

export default startServerAndCreateLambdaHandler(server, requestHandler, {
  middleware: [
    cookieMiddleware,
  ],
});
```

More use-cases and API information can be found in the [library's README](https://github.com/apollo-server-integrations/apollo-server-integration-aws-lambda#middleware).

## Event extensions

In many cases, API Gateway events will have an authorizer in front of them that contains custom state that will be used for authorization during GraphQL resolution. All of the handlers that are packaged with the library contain a generic type which allows you to explicitly extend the base event type. By passing an event with authorization information, that event type will be used during the creation of `contextValue` and for `middleware`. Below is an example, and more information can be found in the [library's README](https://github.com/apollo-server-integrations/apollo-server-integration-aws-lambda/blob/main/README.md#event-extensions).

> 

Reviewer: ```suggestion
```

> ```ts

Reviewer: Should we use `<MultiCodeBlock>` here as well? (and throughout?)

```

</MultiCodeBlock>

> ## Customizing HTTP behavior

Reviewer: Isn't this what middleware allows for now?

I'm not opposed to keeping the vendia stuff in our docs / on this page, but maybe more as a footnote here under a differently named section?

Author: I think there is still value in noting the express integration given that our implementation assumes (without event interception) that the `ApolloServer` is the _only_ thing running on the handler. If a true proxy integration is desired, with multiple route-based handlers, the vendia implementation should be noted.

I think I updated it to this wording elsewhere, but I think "Customizing HTTP Routing Behavior" might be a better fit.
